Metadata reports generator (Issue #3999) (#5531)
dotnet/extensions: src/Packages/Microsoft.Extensions.AuditReports/README.md

- **Metrics**: Reports on the use of source-generated metric definitions in the code.
- **Compliance**: Reports on the use of privacy-sensitive data in the code, including source-generated logging methods.

> - **MetaData** : Is a Combination of both reports above.

Reviewer: ```suggestion
- **Metadata** : Is a Combination of both reports above.
```

The table below shows various MSBuild properties that you can use to control the behavior of the reports generation:

> | Metrics report generator | Compliance report generator | Meta data report generator | Description |

Reviewer: ```suggestion
| Metrics report generator | Compliance report generator | Metadata report generator | Description |
```

]
```

> ## Example of a meta data report

Reviewer: ```suggestion
## Example of a metadata report
```

        /// </summary>
        public const string Target = nameof(Target);

        /// <summary>
        /// The reason for the failure, e.g. the exception message or the HTTP status code.
        /// </summary>
        public const string FailureReason = nameof(FailureReason);
    }

    /// <summary>
    /// The counter metric for the number of failed requests.
    /// </summary>
    [Counter(Tags.Target, Tags.FailureReason)]
    public static partial FailedRequestCounter CreateFailedRequestCounter(Meter meter);
}
```
and let's also assume we have a project with a class that contains privacy-sensitive data:
```csharp
namespace ComplianceTesting
{
    internal sealed class User
    {
        internal User(string name, DateTimeOffset registeredAt)
        {
            Name = name;
            RegisteredAt = registeredAt;
        }

        [PrivateData]
        public string Name { get; }

        public DateTimeOffset RegisteredAt { get; }
    }
}
```

`Microsoft.Extensions.Compliance.Testing` package contains a definition for `[PrivateData]` attribute, we use it here for demonstration purposes only.

> A meta data report for the code listed above might look like this:

Reviewer: ```suggestion
A metadata report for the code listed above might look like this:
```
